### PR TITLE
Shared ptr

### PR DESCRIPTION
**Pull request by NeoBauer**
#Marcinati
#NeoBauer

### CODE REVIEW

File: homework/shared_ptr/SharedPtr.hpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    } else if (ControlBlock_->sharedRefsCounter_ == 1 &&
+               ControlBlock_->weakRefsCounter_ == 0) {
+      ControlBlock_->deleter(rawPtr_);
+      delete ControlBlock_;
+    }
+  }
+}
+
+template <typename T> T *SharedPtr<T>::get() const {
+  if (rawPtr_) {
```
Comment (MaciejSalamonski) [inline homework/shared_ptr/SharedPtr.hpp:69]: Isn't it better to just return rawPtr_?

File: homework/shared_ptr/SharedPtr.hpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    return rawPtr_;
+  } else
+    return nullptr;
+}
+
+template <typename T> T &SharedPtr<T>::operator*() const noexcept {
```
Comment (MaciejSalamonski) [inline homework/shared_ptr/SharedPtr.hpp:75]: I suggest you to throw an exception. We cannot derefenrce a nullptr

**Review by MaciejSalamonski — APPROVED** [review]
Hi! Good job 👍 Maybe implementation of operators= is a bit overcomplicated  but its fine. Have a nice day!

**Comment by J-Cieplinski** [comment]
I'm pretty sure at least some parts of those long `if else` blocks can be put into separate function. DRY

File: homework/shared_ptr/SharedPtr.hpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      delete ControlBlock_;
+      rawPtr_ = other.rawPtr_;
+      ControlBlock_ = other.ControlBlock_;
+      ControlBlock_->sharedRefsCounter_++;
+      return *this;
+    } else
+      return *this;
+  } else
+    return *this;
+}
```
Comment (jakjan95) [inline homework/shared_ptr/SharedPtr.hpp:111]: imo there is too much logic

File: homework/shared_ptr/SharedPtr.hpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+      rawPtr_ = other.rawPtr_;
+      ControlBlock_ = other.ControlBlock_;
+      other.ControlBlock_ = nullptr;
+      other.rawPtr_ = nullptr;
+      return *this;
+    } else
+      return *this;
+  } else
+    return *this;
+}
```
Comment (jakjan95) [inline homework/shared_ptr/SharedPtr.hpp:144]: same here

File: homework/shared_ptr/SharedPtr.hpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+      ControlBlock_ = new ControlBlock<T>{};
+    } else if (ControlBlock_->sharedRefsCounter_ == 1 &&
+               ControlBlock_->weakRefsCounter_ == 0) {
+      ControlBlock_->deleter(rawPtr_);
+      delete ControlBlock_;
+      rawPtr_ = NewRawPtr;
+      ControlBlock_ = new ControlBlock<T>{};
+    }
+  }
+}
```
Comment (jakjan95) [inline homework/shared_ptr/SharedPtr.hpp:170]: same here, you can see some pattern which you can separate do function

**Review by jakjan95 — COMMENTED** [review]
Some comments from me :)

File: homework/shared_ptr/SharedPtr.hpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        }
+    }
+}
+
+template <typename T>
+T* SharedPtr<T>::getPtr() const {
+    if (rawPtr_) {
+        return rawPtr_;
+    } else
+        return nullptr;
```
Comment (WBigaj) [inline homework/shared_ptr/SharedPtr.hpp:77]: return rawPtr_ ? rawPtr_ : nullptr;

File: homework/shared_ptr/SharedPtr.hpp [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            ControlBlock_->deleter(rawPtr_);
+            delete ControlBlock_;
+            rawPtr_ = other.rawPtr_;
+            ControlBlock_ = other.ControlBlock_;
+            ControlBlock_->sharedRefsCounter_++;
+            return *this;
+        } else
+            return *this;
+    } else
+        return *this;
```
Comment (WBigaj) [inline homework/shared_ptr/SharedPtr.hpp:122]: You are returning *this regardless of result.Why do you repeat it everywhere?

File: homework/weak_ptr/WeakPtr.hpp [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+size_t WeakPtr<T>::useSCount() const noexcept {
+    return ControlBlock_->sharedRefsCounter_;
+}
+
+template <typename T>
+bool WeakPtr<T>::expired() const noexcept {
+    if (ControlBlock_) {
+        return useCount() == 0;
+    }
+    return true;
```
Comment (WBigaj) [inline homework/weak_ptr/WeakPtr.hpp:124]:
```suggestion
    return ControlBlock_ ? (useCount() == 0) : true;
```

File: homework/weak_ptr/WeakPtr.hpp [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+template <typename T>
+SharedPtr<T> WeakPtr<T>::lock() const noexcept {
+    if (expired()) {
+        return SharedPtr<T>();
+    }
+    return SharedPtr<T>(*this);
```
Comment (WBigaj) [inline homework/weak_ptr/WeakPtr.hpp:132]:
```suggestion
    return expired() ?  SharedPtr<T>() : SharedPtr<T>(*this);
```

**Review by WBigaj — COMMENTED** [review]
Few suggestions from me.

File: homework/weak_ptr/tests/tests.cpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+constexpr size_t power = 200;
+
+SCENARIO("Testing all WeakPtr functions") {
+    GIVEN("SharedPtr of Scooter with Power = 200") {
+        SharedPtr<Scooter> sharedscooter(new Scooter(power));
+        WHEN("making a WeakPtr from SharedPtr") {
+            WeakPtr<Scooter> weakscooter(sharedscooter);
+            THEN("made a WeakPtr without any problems") {
+                REQUIRE(Catch::isOk);
```
Comment (ziobron) [inline homework/weak_ptr/tests/tests.cpp:13]: you may check if weak ptr is not null

File: homework/weak_ptr/tests/tests.cpp [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+
+    GIVEN("WeakPtr made of SharedPtr of Scooter with Power = 200") {
+        SharedPtr<Scooter> sharedscooter(new Scooter(power));
+        WeakPtr<Scooter> weakscooter(sharedscooter);
+        WHEN("making another WeakPtr from existing first WeakPtr") {
+            WeakPtr<Scooter> weakscootermove(std::move(weakscooter));
+            THEN("the first WeakPtr should expired") {
+                REQUIRE(weakscooter.expired() == true);
```
Comment (ziobron) [inline homework/weak_ptr/tests/tests.cpp:24]: check use count and if pointers are pointing to proper objects or merge it with below test

File: homework/weak_ptr/tests/tests.cpp [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        WHEN("making a WeakPtr from SharedPtr") {
+            WeakPtr<Scooter> weakscooter(sharedscooter);
+            THEN("both shared counter and weak counter should equals 1") {
+                REQUIRE(weakscooter.useCount() == 1);
+                REQUIRE(weakscooter.useSCount() == 1);
+            }
+        }
+    }
+
+    GIVEN("Two SharedPtrs") {
```
Comment (ziobron) [inline homework/weak_ptr/tests/tests.cpp:40]: there are 3 sharedPtrs below

File: homework/weak_ptr/tests/tests.cpp [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        }
+    }
+
+    GIVEN("SharedPtr of Scooter with Power = 200") {
+        SharedPtr<Scooter> sharedscooter(new Scooter(power));
+        WHEN("making a WeakPtr") {
+            WeakPtr<Scooter> weakscooter(sharedscooter);
+            THEN("WeakPtr shouldn't expired") {
+                REQUIRE(weakscooter.expired() == false);
```
Comment (ziobron) [inline homework/weak_ptr/tests/tests.cpp:58]:
```suggestion
                REQUIRE_FALSE(weakscooter.expired());
```

File: homework/weak_ptr/tests/tests.cpp [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        SharedPtr<Scooter> sharedscooter(new Scooter(power));
+        WeakPtr<Scooter> weakscooter(sharedscooter);
+        WHEN("removing the object") {
+            weakscooter.reset();
+            THEN("WeakPtr should expired") {
+                REQUIRE(weakscooter.expired() == true);
+            }
+        }
+    }
+}
```
Comment (ziobron) [inline homework/weak_ptr/tests/tests.cpp:73]: lock() function is not tested.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/shared_ptr/SharedPtr.hpp [hunk 15]
```diff
@@ -4,186 +4,148 @@
...
-            delete ControlBlock_;
+SharedPtr<T>::~SharedPtr()
+{
+    if (controlBlock_ != nullptr)
+    {
+        controlBlock_->sharedRefsCounter_--;
+        if (controlBlock_->sharedRefsCounter_ == 0 && controlBlock_->weakRefsCounter_ == 0)
+        {
+            controlBlock_->deleter(rawPtr_);
+            delete controlBlock_;
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:66]: you release all resources only when there are no shared and weak pointers. std has another behavior. If there are no shared_ptrs resource is deleted (deleter is called) and when there are no shared and weak pointers - control block is removed.

File: homework/shared_ptr/SharedPtr.hpp [hunk 16]
```diff
@@ -4,186 +4,148 @@
...
-    T* getPtr() const;
-    ControlBlock<T>* getControlBlock() const;
-    T& operator*() const noexcept;
-    T* operator->() const noexcept;
-    SharedPtr<T>& operator=(const SharedPtr<T>& other) noexcept;
-    SharedPtr<T>& operator=(SharedPtr<T>&& other) noexcept;
+    T *get() const;
+    T &operator*() const noexcept;
+    T *operator->() const noexcept;
+    SharedPtr<T> &operator=(const SharedPtr<T> &other);
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:26]: please do not commit style changes. It makes code review so much harder.

File: homework/shared_ptr/SharedPtr.hpp [hunk 17]
```diff
@@ -4,186 +4,148 @@
...
+SharedPtr<T> &SharedPtr<T>::operator=(const SharedPtr<T> &other)
+{
+    if (other.rawPtr_ != rawPtr_)
+    {
+        controlBlock_->sharedRefsCounter_--;
+
+        if (controlBlock_->sharedRefsCounter_ == 0 && controlBlock_->weakRefsCounter_ == 0)
+        {
+            controlBlock_->deleter(rawPtr_);
+            delete controlBlock_;
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:99]: duplication (the same code in destructor). Extract to function.

Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:99]: duplication (the same code in destructor). Extract to function.

File: homework/shared_ptr/SharedPtr.hpp [hunk 18]
```diff
@@ -4,186 +4,148 @@
...
-            ControlBlock_ = other.ControlBlock_;
-            ControlBlock_->sharedRefsCounter_++;
-            rawPtr_ = other.rawPtr_;
-            return *this;
-        } else if (ControlBlock_->sharedRefsCounter_ == 1 &&
-                   ControlBlock_->weakRefsCounter_ >= 1) {
-            ControlBlock_->deleter(rawPtr_);
+SharedPtr<T> &SharedPtr<T>::operator=(const SharedPtr<T> &other)
+{
+    if (other.rawPtr_ != rawPtr_)
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:92]:
```suggestion
    if (this != &other)
```

File: homework/shared_ptr/SharedPtr.hpp [hunk 19]
```diff
@@ -4,186 +4,148 @@
...
-            delete ControlBlock_;
-            rawPtr_ = other.rawPtr_;
-            ControlBlock_ = other.ControlBlock_;
-            ControlBlock_->sharedRefsCounter_++;
-            return *this;
-        } else
-            return *this;
-    } else
-        return *this;
+            controlBlock_ = other.controlBlock_;
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:101]: rawPtr_ assignment is missing

File: homework/shared_ptr/SharedPtr.hpp [hunk 20]
```diff
@@ -4,186 +4,148 @@
...
-        return *this;
+SharedPtr<T> &SharedPtr<T>::operator=(SharedPtr<T> &&other) noexcept
+{
+    controlBlock_->sharedRefsCounter_--;
+
+    if (controlBlock_->sharedRefsCounter_ == 0 && controlBlock_->weakRefsCounter_ == 0)
+    {
+        controlBlock_->deleter(rawPtr_);
+        delete controlBlock_;
+        rawPtr_ = other.rawPtr_;
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:117]: where is controlBlock assignment?

File: homework/shared_ptr/SharedPtr.hpp [hunk 21]
```diff
@@ -4,186 +4,148 @@
...
-            delete ControlBlock_;  // Should be managed by weak_Ptr
-            rawPtr_ = NewRawPtr;
-            ControlBlock_ = new ControlBlock<T>{};
-        } else if (ControlBlock_->sharedRefsCounter_ == 1 &&
-                   ControlBlock_->weakRefsCounter_ == 0) {
-            ControlBlock_->deleter(rawPtr_);
-            delete ControlBlock_;
+void SharedPtr<T>::reset(T *NewRawPtr)
+{
+    if (NewRawPtr != rawPtr_)
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:133]: no such check is needed here.

File: homework/weak_ptr/WeakPtr.hpp [hunk 22]
```diff
@@ -26,50 +26,50 @@ class WeakPtr {
...
-        delete ControlBlock_;
+    if (controlBlock_->sharedRefsCounter_ == 0 && controlBlock_->weakRefsCounter_ == 0) {
+        delete controlBlock_;
     }
 }
 
 template <typename T>
 void WeakPtr<T>::weakCounterDecrementer() {
-    ControlBlock_->weakRefsCounter_.exchange(ControlBlock_->weakRefsCounter_.load(std::memory_order_relaxed) - 1,
+    controlBlock_->weakRefsCounter_.exchange(controlBlock_->weakRefsCounter_.load(std::memory_order_relaxed) - 1,
```
Comment (ziobron) [inline homework/weak_ptr/WeakPtr.hpp:41]:
```suggestion
    controlBlock_->weakRefsCounter_--
```
simple decrementation and incrementation works on atomics

File: homework/weak_ptr/WeakPtr.hpp [hunk 23]
```diff
@@ -26,50 +26,50 @@ class WeakPtr {
...
-    otherPtr.ControlBlock_ = nullptr;
+    otherPtr.controlBlock_ = nullptr;
 }
 
 template <typename T>
 WeakPtr<T>::~WeakPtr() {
-    if (ControlBlock_) {
+    if (controlBlock_) {
         weakCounterDecrementer();
         controlBlockRemover();
```
Comment (ziobron) [inline homework/weak_ptr/WeakPtr.hpp]: name your methods with verbs
decrementWeakCounter()
removeControlBlock()

File: homework/weak_ptr/WeakPtr.hpp [hunk 24]
```diff
@@ -99,26 +99,26 @@ template <typename T>
...
 }
 
 template <typename T>
 size_t WeakPtr<T>::useCount() const noexcept {
-    return ControlBlock_->weakRefsCounter_;
+    return controlBlock_->weakRefsCounter_;
 }
 
 template <typename T>
 size_t WeakPtr<T>::useSCount() const noexcept {
```
Comment (ziobron) [inline homework/weak_ptr/WeakPtr.hpp:115]: std::weak_ptr has one function - use_count. It returns a number of shared counter.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/shared_ptr/SharedPtr.hpp [hunk 25]
```diff
@@ -3,149 +3,124 @@
...
-{
+template <typename T> class ControlBlock {
 public:
-    std::atomic<size_t> sharedRefsCounter_{1};
-    std::atomic<size_t> weakRefsCounter_{0};
-    void deleter(T *rawPtr) { delete rawPtr; };
+  std::atomic<size_t> sharedRefsCounter_{1};
+  std::atomic<size_t> weakRefsCounter_{0};
+  void deleter(T *rawPtr) {
+    if (rawPtr != nullptr) {
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:11]: no need for it. calling delete on nullptr is safe.

File: homework/shared_ptr/SharedPtr.hpp (hunk 2 above)
Comment (ziobron) [reply homework/shared_ptr/SharedPtr.hpp:76]: std::shared_ptr does not throw an exception. It's fine and up to developers.

File: homework/shared_ptr/SharedPtr.hpp [hunk 26]
```diff
@@ -3,149 +3,124 @@
...
-            controlBlock_->deleter(rawPtr_);
-            delete controlBlock_;
-            rawPtr_ = NewRawPtr;
-            controlBlock_ = new ControlBlock<T>{};
-        }
+template <typename T> void SharedPtr<T>::reset(T *NewRawPtr) {
+    
+    controlBlock_->sharedRefsCounter_--;
+    deleteAllIfCountersZero();
+    if(controlBlock_ == nullptr){
```
Comment (ziobron) [inline homework/shared_ptr/SharedPtr.hpp:118]: this if is not needed. You should always allocate a new ControlBlock, but if NewRawPtr == nullptr then it should have shared_refcount = 0 not 1.

File: homework/shared_ptr/tests/tests.cpp [hunk 27]
```diff
@@ -1,192 +1,130 @@
...
-                REQUIRE(NewScooter.get()->power_ == power);
-                REQUIRE(scooter.get() == nullptr);
-            }
-        }
+  GIVEN("Two Shared Ptr, one empty, one not") {
+    int power{100};
+    SharedPtr<int> OldSharedPtr(new int(power));
+    SharedPtr<int> NewSharedPtr(nullptr);
+    REQUIRE(OldSharedPtr.use_count() == 1);
+    REQUIRE(NewSharedPtr.use_count() == 1);
```
Comment (ziobron) [inline homework/shared_ptr/tests/tests.cpp:107]:
```suggestion
    REQUIRE(NewSharedPtr.use_count() == 0);
```
AFAIK it should be 0 for nullptr. Please check with std::shared_ptr behavior

File: homework/shared_ptr/tests/tests.cpp [hunk 28]
```diff
@@ -1,192 +1,130 @@
...
+    WHEN("Using move assignment operator from SharedPtr") {
+      NewSharedPtr = std::move(OldSharedPtr);
+      THEN("Switched ownership from OldSharedPtr to NewSharedPtr") {
+        REQUIRE(NewSharedPtr.use_count() == 1);
+        REQUIRE(*(NewSharedPtr.get()) == power);
+        REQUIRE(OldSharedPtr.get() == nullptr);
```
Comment (ziobron) [inline homework/shared_ptr/tests/tests.cpp:113]:
```suggestion
        REQUIRE(OldSharedPtr.get() == nullptr);
        REQUIRE(OldSharedPtr.use_count() == 0);
```

File: homework/shared_ptr/tests/tests.cpp [hunk 29]
```diff
@@ -1,192 +1,130 @@
...
-            }
-        }
+  GIVEN("One SharedPtr and one RawPtr") {
+    int power1{100};
+    int power2{200};
+    SharedPtr<int> SharedOne(new int(power1));
+    int *Power2Ptr = new int(power2);
+    WHEN("Using reset to insert NewScooter") {
+      SharedOne.reset(Power2Ptr);
+      THEN("scooter has power of NewScooter") {
```
Comment (ziobron) [inline homework/shared_ptr/tests/tests.cpp:125]: there is still scooter in description

File: homework/shared_ptr/tests/tests.cpp [hunk 30]
```diff
@@ -1,192 +1,130 @@
...
+        REQUIRE(*(SharedOne.get()) == power2);
```
Comment (ziobron) [inline homework/shared_ptr/tests/tests.cpp:126]: you can also check use_count before and after reset. 
Please also add a test when you use `reset()` (without argument). After such reset use_count should be 0.

**Review by ziobron — CHANGES_REQUESTED** [review]
Can you please fix it by tomorrow end of the day?

**Comment by ziobron** [comment]
Let me grant you only 50 XP for everything
{
  "requirement_ids": [3794],
  "points": [50],
  "users": ["Marcinati", "NeoBauer"]
}

**Comment by ziobron** [comment]
{
"requirement_ids": [1636],
"users": ["Marcinati", "NeoBauer"]
}

**Pull request closed without merging**